### PR TITLE
Private events

### PR DESCRIPTION
**Pull request by patrikis-angelos**
This project is about using advanced associations and the session hash with rails
We created:
- A simple site that you can sign up (without password)
- The user can create an event if he is logged in
- The user can attend an event if he is logged in

### CODE REVIEW

File: app/views/layouts/application.html.erb [hunk 1]
```diff
@@ -11,6 +11,20 @@
...
+    <% if session[:id] == nil%>
+      <%= link_to "Login", users_sign_in_path%>
+    <% else %>
+      <%= "Logged in as: " %>
+      <%= link_to session[:username], "/users/#{session[:id]}"%>
+      <br>
+      <%= link_to "Logout", users_log_out_path, method: :delete%>
+      <br>
+      <%= link_to "Create Event", new_event_path%>
+    <% end %>
```
Comment (oloomoses) [inline app/views/layouts/application.html.erb]: - [x] Move your logic into helpers

File: app/views/users/new.html.erb [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+<% if session[:id] == nil%>
+  <h1>Sign Up</h1>
+  <%= form_with model: @user do |f| %>
+    <%= f.label :username, "Username"%>
+    <%= f.text_field :username, placeholder: "Enter your username"%>
+    <br>
+    <br>
+    <%= f.submit "Submit"%>
+  <% end %>
+<% end %>
```
Comment (oloomoses) [inline app/views/users/new.html.erb]: - [x] move this logic into the helpers

File: app/views/events/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+<br>
+<%= @event.location%>
+<br>
+<%= @event.date%>
+<br>
+<%= @event.description%>
+<br>
+<% if !@event.attendees.exists?(session[:id]) && session[:id] != nil%>
+  <%= link_to "Attend this event", "/events/add/#{@event.id}", method: :put %>
+<% end %>
```
Comment (oloomoses) [inline app/views/events/show.html.erb]: - [x] Move this logic into helpers

**Review by oloomoses — CHANGES_REQUESTED** [review]
# Status: Changes Required
Hello team, great job on this project especially the readme :muscle: 
You are almost there. Just a few things to check on :point_down: 
- [x] Event's index page displays past and upcoming events lists ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from Odin)
- [x] Navigation links are present at the top of the page ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/navigation_links_requirement.png) from Odin)
- [x] Remember to move all your logic into the helpers
##### EventsController

###### `create` action

* [x] uses #build method on association, i.e. `current_user.events.build(event_params)` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/build_method_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

###### `index` action

* [x] lists past events, i.e. `@past_events = Event.previous_events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
* [x] lists ucomping events, i.e. `@upcoming_events = Event.upcoming_events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

##### UsersController

###### `show` action

* [x] lists events created by user, i.e. `@created_events = current_user.events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/current_user_show_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
* [x] lists events previoulsy attendented by user **using event model scope**, i.e. `@past_events = current_user.attended_events.previous` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
* [x] lists upcoming events previously attended by user **using event model scope**, i.e. `@upcoming_events = current_user.attended_events.upcoming` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

Kindly implement the above changes before submitting another review.
### feel free to reach me on slack [Oloo Moses] incase of doubts or concerns

**Comment by descholar-ceo** [comment]
# This Pull Request is under review by [descholar-ceo](https://github.com/descholar-ceo) :eyes: :eyes: 
## Please wait a bit ...

**Review by descholar-ceo — CHANGES_REQUESTED** [review]
# Status: Changes Required :vertical_traffic_light: :warning: 
## Hello @Rayan84 and @patrick-angelos  :handshake: :handshake: :handshake: 

:+1: This is good to make it this far, your project is really nice, congratulations on the good work you have done so far.
:+1: I liked that you have tried to implemented the requested changes from the previous review, keep it up, and your project is really almost done.
#### :warning: However we need to make few changes to mark this project approved, don't feel discouraged because of being asked some changes more and more, that's beacuse this is a learning process, and we are here to help you to become an out-standing developer. See the requested change below :point_down: :point_down: :point_down: 
### Note: All of requested changes which are marked as [Required] are mandatory to be implemented in order to get this project approved, others which are marked as [Optional] are nice to have in your project to make it outstanding

# [Required Changes] :red_circle: 
- [x] You have some logic in views: View-related logic can be encapsulated in View helpers (app/helpers/*_helper.rb)
  - [x] There shoud be absolutely NO logic in views. >Views are only for displaying information that became available to it from its corresponding controller not generating it. [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright _=> To make sure that I clarify what you need to change on this reqirement: Ensure that there is no if conditions, and no loops in your view (`.html.erb`) files_
  
#### Resources to help you learn about this:

- [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code/)
- [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)

# [Optional changes] :yellow_circle:
- [ ] You can add some styin to your page so that it looks way professional 
___
_If anything seems to be unclear to you, reachout and ask anything by commenting on this Pull Request and mention me as [descholar-ceo](https://github.com/descholar-ceo), I will receive a notification and will respond._

[GitHub](https://github.com/descholar-ceo) | [LinkedIn](https://linkedin.com/in/mugirase-emmanuel) | [Twitter](https://twitter.com/descholar3)

_Reviewed by [Mugirase Emmanuel (descholar)](https://github.com/descholar-ceo)_

**Review by mke2111 — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED 🔴 ♻️ 

### Hey @patrick-angelos & @Rayan84 , 👋🏽 
**This is a great project you have going on, well-done lads and thanks for implementing the changes requested earlier on**. However, there are a few issues to fix to get approved.

### REQUIRED CHANGES: ⚠️ 

- [x] Add integration test for all the different features, attend an event, unattend an event, login, signup, create event and so on...
- [x] Test the association between models(this is not hard you can add shoulda gem to make the association test easier).
- [x] Also unit tests to test model validations.

### Kindly make all the requested changes and submit another CRR. 💯 

**Cheers 🍹 and Happy coding 💻** 

## Reach me on slack for a chit-chat 🆘  Roy Mukuye

**Review by MarilenaRoque — APPROVED** [review]
## Hello Team! :rocket:

Congratulations on achieving all the requirements for this project.
Nice implementation of the tests. :tada:

# Your Project is Approved! 🥇 

### Suggention:
- The feature that separates the past and upcoming events at the show page of the user is working just fine, but the title doesn't indicate the upcoming events. It would be nice to divide it into "Past Events" and "Upcoming Events" instead of Past Events and Attended Events.

Keep going with the hard work! 💪🏻 
Happy Coding! 💻 

**_Marilena Roque
Full-Stack Developer
Code Reviewer at Microverse Inc.
[Portfolio](https://mariroqueproject.com/) | [Linkedin](https://www.linkedin.com/in/roquemarilena/) 
marilenas@id.uf.br_**

**Pull request merged**